### PR TITLE
FK & UNQ Constraint Violations

### PR DESCRIPTION
**Pull request by Hydrocharged**
Implements foreign key and unique key constraint violations.
Tests may be found at https://github.com/dolthub/dolt/pull/1870

### CODE REVIEW

File: go/libraries/doltcore/merge/constraint_violations.go [hunk 1]
```diff
@@ -0,0 +1,507 @@
...
+
+		newRoot, err = newRoot.PutTable(ctx, childTblName, postChildTable)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return newRoot, nil
+}
+
+//TODO: stolen from IndexEditor, need to move the original somewhere sensible before merging
```
Comment (bheni) [inline go/libraries/doltcore/merge/constraint_violations.go:153]: Add a method to schema.Index so you can just do index.ToTableTuple(nbf, indexKey) or something?

File: go/libraries/doltcore/merge/constraint_violations.go [hunk 2]
```diff
@@ -0,0 +1,507 @@
...
+
+	"github.com/dolthub/dolt/go/libraries/doltcore/diff"
+	"github.com/dolthub/dolt/go/libraries/doltcore/doltdb"
+	"github.com/dolthub/dolt/go/libraries/doltcore/row"
+	"github.com/dolthub/dolt/go/libraries/doltcore/schema"
+	"github.com/dolthub/dolt/go/libraries/doltcore/table"
+	"github.com/dolthub/dolt/go/libraries/doltcore/table/editor"
+	"github.com/dolthub/dolt/go/libraries/doltcore/table/typed/noms"
+	"github.com/dolthub/dolt/go/store/types"
+)
```
Comment (bheni) [inline go/libraries/doltcore/merge/constraint_violations.go:32]: There are a lot of objects used in this calculation.  For example:

parentFkConstraintViolations(
				ctx,
				foreignKey,
				postChildTable,
				preParentSch, postParentSch, postChildSch,
				postParentIndex, postChildIndex,
				preParentData, postParentData, postChildData, postParentIndexData, postChildIndexData)

Could possibly clean this up and get rid of some code duplication by creating a class that groups some of these elements together.

File: go/libraries/doltcore/merge/constraint_violations.go [hunk 3]
```diff
@@ -0,0 +1,507 @@
...
+			if err != nil {
+				return types.Tuple{}, err
+			}
+		}
+	}
+	return types.NewTuple(format, resVals...)
+}
+
+// parentFkConstraintViolations processes foreign key constraint violations for the parent in a foreign key.
+func parentFkConstraintViolations(
```
Comment (bheni) [inline go/libraries/doltcore/merge/constraint_violations.go:191]: This function is a monster.  Could use some refactoring to improve readability.

File: go/libraries/doltcore/merge/constraint_violations.go [hunk 4]
```diff
@@ -0,0 +1,507 @@
...
+					return tuple.StartsWith(postParentIndexPartialKey), nil
+				}}})
+			//TODO: deal with this defer better
+			defer mapIter.Close(ctx)
+			if _, err := mapIter.ReadRow(ctx); err == nil {
+				// If the parent has more rows that satisfy the partial key then we choose to do nothing
+				continue
+			} else if err != io.EOF {
+				return nil, err
+			}
```
Comment (bheni) [inline go/libraries/doltcore/merge/constraint_violations.go:266]: Can put this all in an anonymous function to fix your defer call

File: go/libraries/doltcore/merge/constraint_violations.go [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+			mapIter = noms.NewNomsRangeReader(postChildIndexSch, postChildIndexData,
+				[]*noms.ReadRange{{Start: postChildIndexPartialKey, Inclusive: true, Reverse: false, Check: func(tuple types.Tuple) (bool, error) {
+					return tuple.StartsWith(postChildIndexPartialKey), nil
+				}}})
+			//TODO: deal with this defer better
+			defer mapIter.Close(ctx)
+			var postChildIndexRow row.Row
+			for postChildIndexRow, err = mapIter.ReadRow(ctx); err == nil; postChildIndexRow, err = mapIter.ReadRow(ctx) {
```
Comment (bheni) [inline go/libraries/doltcore/merge/constraint_violations.go:285]: extract this into a method to fix the defer close issue?  Also this function is a beast so could also help with clarity to pull this out.

**Review by bheni — COMMENTED** [review]
Looks pretty good.  Gonna need lots of testing to validate that the logic is correct.

**Review by andrew-wm-arthur — COMMENTED** [review]
This is still merging the primary row storage using `SessionedTableEditor`, which applies reference options when the rows are merged. 
Reference options must be considered after the primary row storage is merged in order to be applied correctly

**Comment by Hydrocharged** [comment]
> 
> 
> This is still merging the primary row storage using `SessionedTableEditor`, which applies reference options when the rows are merged.
> Reference options must be considered after the primary row storage is merged in order to be applied correctly

Foreign keys are disabled while merging.
https://github.com/dolthub/dolt/blob/master/go/libraries/doltcore/merge/merge.go#L793

**Comment by andrew-wm-arthur** [comment]
> > This is still merging the primary row storage using `SessionedTableEditor`, which applies reference options when the rows are merged.
> > Reference options must be considered after the primary row storage is merged in order to be applied correctly
> 
> Foreign keys are disabled while merging the primary row storage.
> https://github.com/dolthub/dolt/blob/151ca17cb21f782a3bcb392423dcdedc4e10d573/go/libraries/doltcore/merge/merge.go#L792-L794

My mistake, I didn't realize we were already ignoring reference options on merge

**Review by bheni — APPROVED** [review]
LGTM after you merge your bats tests.  Minor comments.

File: go/cmd/dolt/commands/merge.go [hunk 6]
```diff
@@ -137,6 +132,13 @@ func (cmd MergeCmd) Exec(ctx context.Context, commandStr string, args []string,
 				return 1
 			}
 
+			if has, err := root.HasConstraintViolations(ctx); err != nil {
+				verr = errhand.BuildDError("error: failed to get constraint violations").AddCause(err).Build()
+			} else if has {
```
[the remark this answers is not in the document]
Comment (Hydrocharged) [reply go/cmd/dolt/commands/merge.go:137]: Working on that now, will be in a separate PR.

File: go/libraries/doltcore/env/actions/staged.go [hunk 7]
```diff
@@ -106,10 +106,11 @@ func stageTablesNoEnvUpdate(
 		return doltdb.Roots{}, err
 	}
 
-	roots.Staged, err = MoveTablesBetweenRoots(ctx, tbls, roots.Working, roots.Staged)
+	err = checkTablesForConstraintViolations(ctx, tbls, roots.Working)
 	if err != nil {
 		return doltdb.Roots{}, err
 	}
+	roots.Staged, err = MoveTablesBetweenRoots(ctx, tbls, roots.Working, roots.Staged)
```
[the remark this answers is not in the document]
Comment (Hydrocharged) [reply go/libraries/doltcore/env/actions/staged.go:113]: Good catch!

**Pull request merged**